src/transformStateWithClones.js

```diff
@@ -7,7 +7,35 @@
  * @return {Object[]}
  */
 function transformStateWithClones(state, actions) {
-  // write code here
+  const statesArray = [];
+  let updatedState = { ...state };
+
+  for (const action of actions) {
+    updatedState = { ...updatedState };
+
+    switch (action.type) {
+      case 'addProperties':
+        Object.assign(updatedState, action.extraData);
+        statesArray.push(updatedState);
+        break;
+
+      case 'removeProperties':
+        for (const key of action.keysToRemove) {
+          delete updatedState[key];
+        }
+        statesArray.push(updatedState);
+        break;
+
+      case 'clear':
+        for (const key in updatedState) {
+          delete updatedState[key];
+        }
+        statesArray.push(updatedState);
```

Comment:
Issue: The same object reference `updatedState` is being pushed to `statesArray`. This means all elements in `statesArray` will end up being the same, reflecting the final state after all actions. 

Suggestion: Push a clone of `updatedState` to `statesArray` instead. You can achieve this by using `statesArray.push({ ...updatedState });` to ensure each state is a distinct snapshot.